Read links joints position

Thomio-Watanabe opened the pull request.

Thomio-Watanabe committed:
Create Rock output port to write link position

Thomio-Watanabe committed:
Export joints and links positions from gazebo to Rock output ports.
link ports give absolute positions in meters.
joint ports give angles in radians.

Thomio-Watanabe committed:
Use new methods to create component ports.
Define joint ports in orogen file.
Link ports are created dynamically through flatfish configuration file (flat_fish.yml).

doudou commented on tasks/ModelTask.cpp:
```diff
@@ -30,101 +29,102 @@ void ModelTask::setGazeboModel(WorldPtr _world,  ModelPtr _model)
-		link_port_list.push_back( std::make_pair(link_port,*link) );
+        for(NameVector::iterator it = link_names.begin(); it != link_names.end(); ++it)
+        {
+            if( (*it).compare( (*link)->GetName() ) == 0 )
```
One usually uses == for equality

Thomio-Watanabe replied:
I don't get it. Please be more specific.

doudou replied:
It is not natural to use .compare() to compare strings for _equality_ in C++. compare() is used for ordering (returns -1, 0 or 1). Use str1 == str2 to compare for equality.

doudou commented on tasks/ModelTask.cpp:
```diff
@@ -30,101 +29,102 @@ void ModelTask::setGazeboModel(WorldPtr _world,  ModelPtr _model)
-		ports()->addPort( *joint_port );
-		joint_port_list.push_back( std::make_pair(joint_port,*joint) );			
+
+        joints_in.names.push_back( (*joint)->GetName() );
```
Indentation's weird

doudou commented on tasks/ModelTask.cpp:
```diff
@@ -30,101 +29,102 @@ void ModelTask::setGazeboModel(WorldPtr _world,  ModelPtr _model)
+    // The robot configuration YAML file must define the link that will be exported.
+    // Than, these links will be loaded in link_names;
+    link_names = _exported_links.get();
+    setLinkPorts();
```
since you have to call setLinkPorts here, also call setJointPorts. (That's really for consistency reasons).

doudou commented on tasks/ModelTask.cpp:
```diff
@@ -30,101 +29,102 @@ void ModelTask::setGazeboModel(WorldPtr _world,  ModelPtr _model)
+
+bool ModelTask::configureHook()
+{
+    if( ! ModelTaskBase::configureHook() )
```
You should verify that setGazeboModel has been called (i.e. the gazebo model is not NULL)

doudou commented on tasks/ModelTask.cpp:
```diff
@@ -30,101 +29,102 @@ void ModelTask::setGazeboModel(WorldPtr _world,  ModelPtr _model)
 } 
-//======================================================================================
+
 void ModelTask::setJointPorts()
```
setJointPorts really does not describe what the method does. Better name: setupJoints

doudou commented on tasks/ModelTask.cpp:
```diff
@@ -30,101 +29,102 @@ void ModelTask::setGazeboModel(WorldPtr _world,  ModelPtr _model)
+                    "/" + (*link)->GetName() << std::endl;
+
+                // Create the ports dynamicaly
+                link_out_port = new RBSOutPort( "link_"+ (*link)->GetName() +"_samples" );
```
Obvious error here.... Let you figure it out.

Thomio-Watanabe replied:
port name ?

doudou commented on tasks/ModelTask.cpp:
```diff
@@ -30,101 +29,102 @@ void ModelTask::setGazeboModel(WorldPtr _world,  ModelPtr _model)
+    {
+        // Apply effort to joint
+        double effort = joints_in.getElementByName( (*it)->GetName() ).effort;
+        if( effort != RTT::NoData )
```
Think very hard about that one ...

doudou commented on tasks/ModelTask.cpp:
```diff
@@ -30,101 +29,102 @@ void ModelTask::setGazeboModel(WorldPtr _world,  ModelPtr _model)
+
+        // Read joint angle from gazebo link
+        double angle = (*it)->GetAngle(0).Radian();
+        joints_out.getElementByName( (*it)->GetName() ).Position(angle);
```
And about that one. Have you tested the code ? Looked at the results ?

Thomio-Watanabe replied:
Just tested, it was broken. I've already fixed (next commit).

doudou commented on tasks/ModelTask.cpp:
```diff
@@ -30,101 +29,102 @@ void ModelTask::setGazeboModel(WorldPtr _world,  ModelPtr _model)
+    {
+        math::Pose link_pose = (*it).second->GetWorldPose();
+        rock_rbs.position = base::Vector3d(link_pose.pos.x,link_pose.pos.y,link_pose.pos.z);
+        rock_rbs.orientation = base::Quaterniond(
```
Have you double-checked the frame convention in gazebo ? Do they match [the one from rock](http://www.ros.org/reps/rep-0103.html) ? (This is really an open question, I do not know the answer)

Thomio-Watanabe replied:
It seems they are equal: "right-handed coordinate system, with x and y in the plane, and z increasing with altitude".
http://playerstage.sourceforge.net/doc/Gazebo-manual-0.8.0-pre1-html/config_syntax.html

doudou commented on tasks/ModelTask.cpp:
```diff
@@ -30,101 +29,102 @@ void ModelTask::setGazeboModel(WorldPtr _world,  ModelPtr _model)
-			link->AddRelativeForce( math::Vector3(force(0),force(1),force(2)) );
-		}
-	}
+    base::samples::RigidBodyState rock_rbs;
```
Where do declarations go, again ?

doudou commented on tasks/ModelTask.cpp:
```diff
@@ -30,101 +29,102 @@ void ModelTask::setGazeboModel(WorldPtr _world,  ModelPtr _model)
+    for(LinkOutput::iterator it = link_list.begin(); it != link_list.end(); ++it)
+    {
+        math::Pose link_pose = (*it).second->GetWorldPose();
+        rock_rbs.position = base::Vector3d(link_pose.pos.x,link_pose.pos.y,link_pose.pos.z);
```
You don't set the sourceFrame and targetFrame for the RBS. This is really critical information.

doudou commented on tasks/ModelTask.cpp:
```diff
@@ -30,101 +29,102 @@ void ModelTask::setGazeboModel(WorldPtr _world,  ModelPtr _model)
+    base::samples::RigidBodyState rock_rbs;
+    for(LinkOutput::iterator it = link_list.begin(); it != link_list.end(); ++it)
+    {
+        math::Pose link_pose = (*it).second->GetWorldPose();
```
I would rather have the RBS be from whatever root the link have to whatever target the link have (i.e. relative pose instead of global pose). At least for now.

Thomio-Watanabe replied:
gazebo::physics::Entity::GetRelativePose() gets the pose of the entity relative to its parent.... 
So, to generalize the target I should take the difference from the source and target absolute positions, right?
How do you think we should define the source/targets ? Through properties ?  Instead of exported_links we use exported_sources/exported_targets ?

doudou replied:
Better define a LinkExport struct in the orogen with:
- source and target links
- source and target frames (the values that should be filled in the RBS)

doudou commented on tasks/ModelTask.cpp:
```diff
@@ -30,101 +29,102 @@ void ModelTask::setGazeboModel(WorldPtr _world,  ModelPtr _model)
 
-	for(LinkPort_V::iterator it = link_port_list.begin(); it != link_port_list.end(); ++it)
-		delete it->first; 
+    joints_in.clear();
```
Do. Not. Clear. Vectors. Explicitely. In. Destructors.

I've been trying to explain why to you three times already. Was I not clear enough ? Do you understand why I say that ?

doudou commented on tasks/ModelTask.hpp:
```diff
@@ -32,24 +33,28 @@ namespace rock_gazebo {
+            base::samples::Joints joints_in, joints_out;
+
+            typedef std::vector<std::string> NameVector;
+            NameVector link_names;
```
You don't need to store this in the class. It is used only once. Make it a local variable.

Thomio-Watanabe committed:
cleanup part-1:
 - Fix indentation (delete tab chars)
 - change mathod names (setJointPorts/setLinkPorts -> setupJoints/setupLinks)
 - check it effort is NaN
 - set sourceFrame and targetFrame
 - verify if model is not NULL
 - call setupLinks() inside configureHook()
 - stop cleaning vectors
 - make link_names a local variable

doudou commented on tasks/ModelTask.cpp:
```diff
@@ -4,127 +4,138 @@
+    {
+        // Apply effort to joint
+        double effort = joints_in.getElementByName( (*it)->GetName() ).effort;
+        if( ! isnan(effort) )
```
Use hasPosition on JointState for this. Moreover, don't test for NaN directly in Rock. Use one of the specific helpers (http://rock.opendfki.de/wiki/WikiStart/Standards/RG3)

Thomio-Watanabe replied:
Do you mean hasEffort() ?!
I cannot access the dfki link above, I don't have permission.

doudou replied:
> Do you mean hasEffort() ?!

Oups. Yes.

You can access the http:// version of it (without the 's') :P Forgot to remove it.

Thomio-Watanabe committed:
cleanup part-2:
  - use hasEffort() to test if the input field is set
  - rename link ports
  - fix joint output port

Thomio-Watanabe committed:
Define LinkExport struct
 ... to set source/target links and calculate their relative position

Thomio-Watanabe committed:
Use LinkExport as exported_links type
Calculate source_link pose relative to target_link

doudou commented:
I can't see anything major left ... merged !
